cli: split the notions of "interactive session" and "terminal output"

knz opened the pull request:
Fixes #19775.
Informs https://github.com/cockroachdb/docs/issues/2263 & https://github.com/cockroachdb/docs/issues/2248

Prior to this patch, a single boolean notion "interactive session" was
used to decide default settings for client commands that issue SQL
queries. This characteristic was decided based on whether both standard
input and standard output are on a terminal.

This characterisation was erroneous however as it didn't properly
account for the following:

- sessions started with `cockroach sql -e`, or non-interactive
  commands like `cockroach node`. These are definitely
  non-interactive, even if the standard input is a terminal.
- sessions started with `cockroach sql >foo.out`, i.e. letting the
  human enter commands from standard input but redirecting output to a
  file. These are arguably interactive and should benefit from the
  various defaults meant to help human users.

To achieve this, this patch separates the notions:

- an *interactive session* is one where the `cockroach` command is
  confident there's a user *manually entering queries*, looking at
  results in-between inputs. This is only decided to be true if 1) the
  `cockroach sql` command is used, 2) the standard input is a terminal
  3) `-e` is not used.
- a session with *terminal output* is where the `cockroach` command
  is confident there's a user *looking at the output*.

These characteristics are then used as follows to decide defaults:

| Setting                        | New default                                             | Previous bug(s) |
|--------------------------------|---------------------------------------------------------|------------------|
| `sql_safe_updates` (session)   | true when the session is interactive and `--unsafe-updates` not given, false otherwise. | defaulted to true always, not only when interactive |
| `check_syntax` (client)        | true when the session is interactive, false otherwise.  | erroneously false when inputting from user with output redirected |
| `errexit` (client)             | false when the session is interactive, true otherwise.  | erroneously true when inputting from user with output redirected |
| `display_format` (client)      | `pretty` when the output goes to a terminal and `--display-format` not given, `tsv` otherwise. | erroneously `tsv` when outputing to terminal with input redirected |
| `show_times` (client)          | true when the output goes to a terminal and `display_format` is `pretty`, false otherwise.     | erroneously false when outputting to terminal with input redirected |
| Input line editor              | true when the session is both interactive and outputting on a terminal, false otherwise.       | N/A |
| `smart_prompt` (client)        | true when the input line editor is enabled.             | N/A |
| Display of input prompt        | true when the input line editor is enabled.             | N/A |
| Storing the input history      | true when the input line editor is enabled.             | N/A |
| `echo` (client)                | always false unless `--echo-sql` given.                 | N/A |

Release note (bug fix, cli change):

- The session variable `sql_safe_updates` defaults to `false` unless
  the shell is truly interactive (using `cockroach sql`, `-e` not
  provided, standard input not redirected) and `--unsafe-updates` is
  not provided.
  Previously it would always default to `true` unless `--unsafe-updates`
  was set.

- The client-side option `errexit` defaults to `false` only if the
  shell is truly interactive, not only when the input is not
  redirected as previously.

- The client-side option `display_format` defaults to `pretty` in
  every case where the output goes to a terminal, not only when the
  input is not redirected as previously.

- `check_syntax`, `smart_prompt` (client-side options), together with
  the interactive line editor, are only enabled if the session is
  interactive and output goes to a terminal.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/20805)
<!-- Reviewable:end -->

knz commented:
Request to cherry-pick this into 1.1.

knz committed:
cli: split the notions of "interactive session" and "terminal output"

Prior to this patch, a single boolean notion "interactive session" was
used to decide default settings for client commands that issue SQL
queries. This characteristic was decided based on whether both standard
input and standard output are on a terminal.

This characterisation was erroneous however as it didn't properly
account for the following:

- sessions started with `cockroach sql -e`, or non-interactive
  commands like `cockroach node`. These are definitely
  non-interactive, even if the standard input is a terminal.
- sessions started with `cockroach sql >foo.out`, i.e. letting the
  human enter commands from standard input but redirecting output to a
  file. These are arguably interactive and should benefit from the
  various defaults meant to help human users.

To achieve this, this patch separates the notions:

- an *interactive session* is one where the `cockroach` command is
  confident there's a user *manually entering queries*, looking at
  results in-between inputs. This is only decided to be true if 1) the
  `cockroach sql` command is used, 2) the standard input is a terminal
  3) `-e` is not used.
- a session with *terminal output* is where the `cockroach` command
  is confident there's a user *looking at the output*.

These characteristics are then used as follows to decide defaults:

| Setting                        | New default                                             | Previous bug(s) |
|--------------------------------|---------------------------------------------------------|------------------|
| `sql_safe_updates` (session)   | true when the session is interactive and `--unsafe-updates` not given, false otherwise. | defaulted to true always, not only when interactive |
| `check_syntax` (client)        | true when the session is interactive, false otherwise.  | erroneously false when inputting from user with output redirected |
| `errexit` (client)             | false when the session is interactive, true otherwise.  | erroneously true when inputting from user with output redirected |
| `display_format` (client)      | `pretty` when the output goes to a terminal and `--display-format` not given, `tsv` otherwise. | erroneously `tsv` when outputing to terminal with input redirected |
| `show_times` (client)          | true when the output goes to a terminal and `display_format` is `pretty`, false otherwise.     | erroneously false when outputting to terminal with input redirected |
| Input line editor              | true when the session is both interactive and outputting on a terminal, false otherwise.       | N/A |
| `smart_prompt` (client)        | true when the input line editor is enabled.             | N/A |
| Display of input prompt        | true when the input line editor is enabled.             | N/A |
| Storing the input history      | true when the input line editor is enabled.             | N/A |
| `echo` (client)                | always false unless `--echo-sql` given.                 | N/A |

Release note (bug fix, cli change):

- The session variable `sql_safe_updates` defaults to `false` unless
  the shell is truly interactive (using `cockroach sql`, `-e` not
  provided, standard input not redirected) and `--unsafe-updates` is
  not provided.
  Previously it would always default to `true` unless `--unsafe-updates`
  was set.

- The client-side option `errexit` defaults to `false` only if the
  shell is truly interactive, not only when the input is not
  redirected as previously.

- The client-side option `display_format` defaults to `pretty` in
  every case where the output goes to a terminal, not only when the
  input is not redirected as previously.

- `check_syntax`, `smart_prompt` (client-side options), together with
  the interactive line editor, are only enabled if the session is
  interactive and output goes to a terminal.

jordanlewis commented on pkg/cli/sql.go:
```diff
@@ -647,7 +658,7 @@ func (c *cliState) doStartLine(nextState cliStateEnum) cliStateEnum {
 	c.partialStmtsLen = 0
 
-	if isInteractive {
+	if c.ins != noLineEditor {
```
nit: this read clearer before - could you add a bool for this with a descriptive name? like hasLineEditor.

jordanlewis reviewed (approved):
LGTM!

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>, but -1 on cherry-picking to 1.1.

---

Review status: 0 of 10 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20805#-:-L0euCKG0j8MyQk4TuFV:blw0a5x)*
<!-- Sent from Reviewable.io -->

knz commented:
TFYRs!
